### PR TITLE
Add support for varmap

### LINKED ISSUE #45: vars should be able to take non-flat values
### What happened?
Usecase: We are trying to create EKS Cluster using Crossplane and to create multiple Node-groups/launch-templates we are using terraform-provider to dynamically read the details from the claim.yaml

claim.yaml

```apiVersion: abc.io/v1alpha1
kind: ManagedCluster
metadata:
  namespace: crossplane-system
  name: eks-cluster
parameters:
    region: us-east-2
    version: "1.21"
    nodeRole: arn:aws:iam::xxxxx:role/eksNodeRole
    nodeGroups:
      - desiredSize: 2
        maxSize: 2
        minSize: 2
        instanceType: m5.4xlarge
        label: fruit=apple
        volumeSize: 150
      - desiredSize: 1
        maxSize: 1
        minSize: 1
        instanceType: m5.4xlarge
        label: fruit=banana
        volumeSize: 80
```

As per current behavior, terraform-provider gives 2 options to pass variables - `vars` & `varFiles`. 

`varFiles` supports data in HCL or JSON format.
`vars` support data in plain string - flat values only.

### What environment did it happen in?
Crossplane version: [1.10.1](https://github.com/crossplane/crossplane/releases/tag/v1.10.1)
provider-terraform: [0.2.0](https://marketplace.upbound.io/providers/upbound/provider-terraform/v0.2.0)
Kubernetes: 1.23

Solution could be to enhance `vars` to take non-flat values as input & internally convert to HCL/JSON to pass to terraform-provider.

### PR DESCRIPTION
**Pull request by bdwyertech**
### Description of your changes

Fixes #45 

I have:

- [ ] Run `make reviewable` to ensure this PR is ready for review.

### CODE REVIEW

**Comment by bobh66** [comment]
@bdwyertech the CI config has been updated, rebasing to the current main should resolve the lint timeout error.

**Comment by bdwyertech** [comment]
> @bdwyertech the CI config has been updated, rebasing to the current main should resolve the lint timeout error.

Should be good to go @bobh66

**Comment by nakamume** [comment]
@bdwyertech thanks for the change - this would greatly improve our workflows!

@bobh66 when would this change be released?

File: internal/controller/workspace/workspace.go [hunk 1]
```diff
@@ -420,6 +421,14 @@ func (c *external) options(ctx context.Context, p v1beta1.WorkspaceParameters) (
...
 	}
 
+	if p.VarMap != nil {
+		jsonBytes, err := json.Marshal(p.VarMap)
+		if err != nil {
+			return nil, errors.Wrap(err, errVarFile)
+		}
+		o = append(o, terraform.WithVarFile(jsonBytes, terraform.JSON))
+	}
+
```
Comment (ytsarev) [inline internal/controller/workspace/workspace.go:431]: Can we please write a test for that in https://github.com/upbound/provider-terraform/blob/main/internal/controller/workspace/workspace_test.go ?

**Review by ytsarev — CHANGES_REQUESTED** [review]
@bdwyertech thank you so much for your contribution, it looks awesome!

As @bobh66 mentioned could you please add an example to https://github.com/upbound/provider-terraform/tree/main/examples and the test for VarMap as in my inline comment?

**Comment by nakamume** [comment]
@ytsarev @bdwyertech We are kinda blocked on our Crossplane journey because of this feature. I'd like continue with these changes. I guess, I'll have to fork and open a new PR. Please let me know if there's some way to contribute to this same PR.

**Comment by ytsarev** [comment]
@nakamume I'm totally open to accepting these changes, we just need to invest in quality and an automated test suite to avoid potential regressions in the future.

**Comment by nakamume** [comment]
Let me open a new PR then with test coverage

**Comment by nakamume** [comment]
@ytsarev opened https://github.com/upbound/provider-terraform/pull/112

**Comment by ytsarev** [comment]
closing in favor of https://github.com/upbound/provider-terraform/pull/112 , thanks a lot for the original implementation! 👍

**Pull request closed without merging**